env/env_encrypt2.cc

```diff
@@ -0,0 +1,989 @@
+  } else {
+    crypto_ = std::make_shared<UnixLibCrypto>();
+    crypto_shared = crypto_;
+  }
+
+  valid_ = crypto_->IsValid();
+  if (IsValid()) {
+    crypto_->RAND_poll();
+  }
+}
+
+void EncryptedEnvV2::SetKeys(EncryptedEnvV2::ReadKeys encrypt_read,
+                             EncryptedEnvV2::WriteKey encrypt_write) {
+  WriteLock lock(&key_lock);
+  encrypt_read_ = encrypt_read;
+  encrypt_write_ = encrypt_write;
+}
+
+bool EncryptedEnvV2::IsWriteEncrypted() const {
+  ReadLock lock(&key_lock);
+  bool ret_flag = (nullptr != encrypt_write_.second);
+  return ret_flag;
+}
+
+//
+// common functions used with different file types
+//  (because there is not common base class for the file types
+//
+template <class TypeFile>
+Status EncryptedEnvV2::ReadSeqEncryptionPrefix(
+    TypeFile* f, std::shared_ptr<const CTREncryptionProviderV2>& provider,
+    std::unique_ptr<BlockAccessCipherStream>& stream) {
+  Status status;
+
+  provider.reset();  // nullptr for provider implies "no encryption"
+  stream.release();
+
+  // Look for encryption marker
+  EncryptMarker marker;
+  Slice marker_slice;
+  status = f->Read(sizeof(marker), &marker_slice, marker);
+  if (status.ok()) {
+    if (sizeof(marker) == marker_slice.size() &&
+        marker_slice.starts_with(kEncryptMarker)) {
+      // code_version currently unused
+      uint8_t code_version = (uint8_t)marker_slice[7];
+
+      if (kEncryptCodeVersion0 == code_version) {
+        Slice prefix_slice;
+        PrefixVersion0 prefix_buffer;
+        status = f->Read(sizeof(PrefixVersion0), &prefix_slice,
+                         (char*)&prefix_buffer);
+        if (status.ok() && sizeof(PrefixVersion0) == prefix_slice.size()) {
+          Sha1Description desc(prefix_buffer.key_description_,
+                               sizeof(prefix_buffer.key_description_));
+
+          ReadLock lock(&key_lock);
+          auto it = encrypt_read_.find(desc);
+          if (encrypt_read_.end() != it) {
+            provider = it->second;
+            stream.reset(new AESBlockAccessCipherStream(
+                provider->key(), code_version, prefix_buffer.nonce_));
+
+          } else {
+            status = Status::NotSupported(
+                "No encryption key found to match input file");
+          }
+        }
+      } else {
+        status =
+            Status::NotSupported("Unknown encryption code version required.");
+      }
+    }
+  }
+  return status;
+}
+
+template <class TypeFile>
+Status EncryptedEnvV2::ReadRandEncryptionPrefix(
+    TypeFile* f, std::shared_ptr<const CTREncryptionProviderV2>& provider,
+    std::unique_ptr<BlockAccessCipherStream>& stream) {
+  Status status;
+
+  provider.reset();  // nullptr for provider implies "no encryption"
+  stream.release();
+
+  // Look for encryption marker
+  EncryptMarker marker;
+  Slice marker_slice;
+  status = f->Read(0, sizeof(marker), &marker_slice, marker);
+  if (status.ok()) {
+    if (sizeof(marker) == marker_slice.size() &&
```

Comment:
This is expected:  unencrypted file detected therefore provider is NOT set, but file is not in error.  Stated otherwise, file did not create an error and happens to not be encrypted.  This allows transition to and from encryption.